net/ipsec/ipsec.go

```diff
@@ -0,0 +1,402 @@
+}
+
+func (ipsec *IPSec) removeMarkRule(srcIP, dstIP net.IP, dstPort int) error {
+	rulespec := markRulespec(srcIP, dstIP, dstPort)
+	if err := ipsec.ipt.Delete(table, mainChain, rulespec...); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("iptables delete (%s, %s, %s)", table, mainChain, rulespec))
+	}
+
+	return nil
+}
+
+func markRulespec(srcIP, dstIP net.IP, dstPort int) []string {
+	return []string{
+		"-s", srcIP.String(), "-d", dstIP.String(),
+		"-p", "udp", "--dport", strconv.FormatUint(uint64(dstPort), 10),
+		"-j", markChain,
+	}
+
+}
+
+func (ipsec *IPSec) resetIPTables(destroy bool) error {
+	if err := ipsec.ipt.ClearChain(table, mainChain); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("iptables clear (%s, %s)", table, mainChain))
+	}
+
+	if err := ipsec.ipt.ClearChain(table, markChain); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("iptables clear (%s, %s)", table, markChain))
+	}
+
+	if err := ipsec.ipt.AppendUnique(table, "OUTPUT", "-j", mainChain); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("iptables append (%s, %s)", table, "OUTPUT"))
+	}
+
+	if !destroy {
+		rulespec := []string{"-j", "MARK", "--set-xmark", markStr}
+		if err := ipsec.ipt.Append(table, markChain, rulespec...); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("iptables append (%s, %s, %s)", table, markChain, rulespec))
+		}
+
+		return nil
+	}
+
+	if err := ipsec.ipt.Delete(table, "OUTPUT", "-j", mainChain); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("iptables delete (%s, %s)", table, "OUTPUT"))
+	}
+
+	if err := ipsec.ipt.DeleteChain(table, mainChain); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("iptables delete (%s, %s)", table, mainChain))
+	}
+
+	if err := ipsec.ipt.DeleteChain(table, markChain); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("iptables delete (%s, %s)", table, mainChain))
+	}
+
+	return nil
+}
+
+// xfrm
+
+func xfrmState(srcIP, dstIP net.IP, spi SPI, key []byte) (*netlink.XfrmState, error) {
+	if len(key) != 36 {
+		return nil, fmt.Errorf("key should be 36 bytes long")
+	}
+
+	return &netlink.XfrmState{
+		Src:   srcIP,
+		Dst:   dstIP,
+		Proto: netlink.XFRM_PROTO_ESP,
+		Mode:  netlink.XFRM_MODE_TRANSPORT,
+		Spi:   int(spi),
+		Aead: &netlink.XfrmStateAlgo{
+			Name:   "rfc4106(gcm(aes))",
+			Key:    key,
+			ICVLen: 128,
+		},
+	}, nil
+}
+
+func xfrmPolicy(srcIP, dstIP net.IP, spi SPI) *netlink.XfrmPolicy {
+	ipMask := []byte{0xff, 0xff, 0xff, 0xff} // /32
+
+	return &netlink.XfrmPolicy{
+		Src:   &net.IPNet{IP: srcIP, Mask: ipMask},
+		Dst:   &net.IPNet{IP: dstIP, Mask: ipMask},
+		Proto: syscall.IPPROTO_UDP,
+		Dir:   netlink.XFRM_DIR_OUT,
+		Mark: &netlink.XfrmMark{
+			Value: mark,
+			Mask:  mark,
+		},
+		Tmpls: []netlink.XfrmPolicyTmpl{
+			{
+				Src:   srcIP,
+				Dst:   dstIP,
+				Proto: netlink.XFRM_PROTO_ESP,
+				Mode:  netlink.XFRM_MODE_TRANSPORT,
+				Spi:   int(spi),
+			},
+		},
+	}
+}
+
+// Helpers
+
+func newSPI(srcPeer, dstPeer mesh.PeerShortID) (SPI, error) {
```

Comment:
> The SPI collisions are harmful only on the recv path

Looks like this is enforced early - I just tried to manually create two SAs with the same dst + SPI and received `RTNETLINK answers: File exists` for the second (please confirm this!). This is actually not so bad - in the unlikely event that the circumstance I described arises, the second `Protect` call will return an error and that connection will fall back to sleeve?

> but such approach requires to exchange some handshake message between peers.

Ah yes of course, because you'd need to communicate the SPI you had allocated locally to the other end.

> Alternatively, we could use a 32-bit hash function to derive SPIs, but that does not prevent from collisions.

True. It looks like collisions are fail-safe though (assuming what I've said above is true) - in which case, perhaps we should pick the method that is least likely to produce collisions?